### PR TITLE
Add RegexFullMatch operator

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Description
This PR introduces the `RegexFullMatch` operator, as originally proposed in https://github.com/onnx/onnx/issues/5317. 

The `RegexFullMatch` operator takes one string tensor as input and returns a bool tensor of identical shape indicating if each element fully matches the regex pattern encoded in the `pattern` string attribute. This attribute is a string and we expect valid [re2](https://github.com/google/re2) regex.

Some examples are as follows:
```
RegexFullMatch(["www.google.com", "www.facebook.com", "www.bbc.co.uk"], "www\.[\w.-]+\.\bcom\b")
=> [True, True, False]
RegexFullMatch([["account@gmail.com", "account@hotmail.com"], ["not email", "account2@yahoo.com"]], "(\W|^)[\w.\-]{0,25}@(yahoo|gmail)\.com(\W|$)")
=> [[True, False], [False, True]]
```

### Motivation and Context
Closes https://github.com/onnx/onnx/issues/5317.

Following discussion at the last Operators SIG Weekly the "engine" attribute has been dropped in favour of simply using [re2](https://github.com/google/re2) syntax for now. This reflects the fact that both [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/strings/regex_full_match) and [PyTorch](https://pytorch.org/text/0.15.0/transforms.html#regextokenizer) operators requiring regex use re2 already.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 numpy
 protobuf>=3.20.2
+google-re2
```
Comment (xadupre) [inline requirements.txt:3]: I would make this dependency optional: `pip install onnx[text]`.

File: onnx/defs/text/defs.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#include "onnx/defs/schema.h"
```
Comment (xadupre) [inline onnx/defs/text/defs.cc:1]: There should be a file `Operators-text.md` as there is a file `Operators-ml.md`. Maybe that can be done separately.

Comment (adityagoel4512) [reply onnx/defs/text/defs.cc:1]: Are you proposing a new domain for text-based operators out of curiosity?

Comment (xadupre) [reply onnx/defs/text/defs.cc:1]: That's what my comments entails but you're right, I remember we only decided to move the text operators in a separate folder and not in a separate domain. Forget my comment. Maybe this question will come again if we keep adding more text operators.

File: requirements.txt (hunk 1 above)
Comment (adityagoel4512) [reply requirements.txt]: Have done so.

**Comment by adityagoel4512** [comment]
Would it be possible to receive a review/approval on this PR please? @gramalingam @xadupre

**Review by jcwchen — COMMENTED** [review]
I just added run-release-CIs label to run release CIs (it will take effect if there is a new commit). You will need to add `google-re2` in `requirements-release.txt` to make release CIs pass I believe.

File: onnx/backend/test/case/node/regex_full_match.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+
+import onnx
+from onnx.backend.test.case.base import Base
+from onnx.backend.test.case.node import expect
+
+
+class RegexFullMatch(Base):
```
Comment (xadupre) [inline onnx/backend/test/case/node/regex_full_match.py:12]: We should add a test with there is no match at all, a null dimension. Do we assume the regular expression is always right?

File: onnx/defs/text/defs.cc [hunk 4]
```diff
@@ -33,6 +29,34 @@ ONNX_OPERATOR_SET_SCHEMA(
                 *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape());
         }));
 
+static const char* RegexFullMatch_doc =
+    R"DOC(RegexFullMatch performs a full regex match on each element of the input element. If an element fully matches the regex pattern specified as an attribute, the corresponding element in the output is True and it is False otherwise. [RE2](https://github.com/google/re2/wiki/Syntax) regex syntax is used.)DOC";
```
Comment (xadupre) [inline onnx/defs/text/defs.cc:33]: input tensor?

File: onnx/reference/ops/op_regex_full_match.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import numpy as np
+import re2
+
+from onnx.reference.op_run import OpRun
+
+_acceptable_str_dtypes = ("U", "O")
+
+
+class RegexFullMatch(OpRun):
+    def _run(self, x, pattern):
```
Comment (xadupre) [inline onnx/reference/ops/op_regex_full_match.py:15]: Replace by `pattern=None` otherwise the reference implementation considers it as an input.

File: requirements-release.txt [hunk 6]
```diff
@@ -7,3 +7,4 @@ wheel
 setuptools
 twine
 parameterized
+google-re2
```
Comment (xadupre) [inline requirements-release.txt:10]: This should be moved to requirements.txt with `google-re2[text]`. See https://pip.pypa.io/en/stable/reference/requirement-specifiers/#overview.

File: requirements-dev.txt [hunk 7]
```diff
@@ -8,6 +8,7 @@ wheel
 jupyter
 setuptools
 twine
+google-re2
```
Comment (xadupre) [inline requirements-dev.txt:11]: This should be moved to requirements.txt with `google-re2[text]`. See https://pip.pypa.io/en/stable/reference/requirement-specifiers/#overview.

File: setup.py [hunk 8]
```diff
@@ -341,6 +341,7 @@ def build_extensions(self):
 with open(requirements_path) as f:
     install_requires = f.read().splitlines()
 
+extras_require["text"] = ["google-re2"]
```
Comment (xadupre) [inline setup.py:344]: It should not be necessary anymore.

File: onnx/backend/test/case/node/regex_full_match.py (hunk 3 above)
Comment (adityagoel4512) [reply onnx/backend/test/case/node/regex_full_match.py:12]: Behaviour should be undefined if an invalid pattern is provided. A reasonable approach may be to raise (there seems to be precedent for this in [Squeeze](https://github.com/onnx/onnx/blob/main/docs/Operators.md#squeeze)) but I'd prefer to let implementations decide.

File: onnx/defs/text/defs.cc [hunk 9]
```diff
@@ -33,6 +29,34 @@ ONNX_OPERATOR_SET_SCHEMA(
                 *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape());
         }));
 
+static const char* RegexFullMatch_doc =
+    R"DOC(RegexFullMatch performs a full regex match on each element of the input tensor. If an element fully matches the regex pattern specified as an attribute, the corresponding element in the output is True and it is False otherwise. [RE2](https://github.com/google/re2/wiki/Syntax) regex syntax is used.)DOC";
```
Comment (gramalingam) [inline onnx/defs/text/defs.cc:33]: I wonder what we should do about changes in RE2 over time? Should we specify a specific version of RE2? In the future, if it evolves, we could potentially add a version attribute to keep evolving.

File: onnx/backend/test/case/node/regex_full_match.py (hunk 3 above)
Comment (gramalingam) [reply onnx/backend/test/case/node/regex_full_match.py:12]: It seems fine for the op spec to clarify that the pattern is expected to be valid. (Then, the behavior will be implementation-dependent.) For the reference-implementation, it seems a good idea to raise an exception with a clear message.

**Comment by adityagoel4512** [comment]
Slight issue here: google-re2 and pyre2 don't ship win32 wheels and one of these needs to be a dependency for the reference implementation of RegexFullMatch (it is not used anywhere else). This is causing the windows release CI to fail as it attempts to build pyre2 from the source distribution which is non trivial.

I had a [look](https://github.com/adityagoel4512/pyre2/actions/runs/5680016161/job/15393282432#step:6:293) at it and it should in principle be possible to build win32 wheels. Before going any further I wanted to first check to see if win32 is actively being supported by ONNX or if it is unintentionally present in this CI? It looks like 32 bit Linux is dropped so I was not sure.

Should I create a separate issue for this?

File: onnx/reference/ops/op_regex_full_match.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) ONNX Project Contributors
+
+# SPDX-License-Identifier: Apache-2.0
+# pylint: disable=R0912,R0913,W0221
+
+import numpy as np
+import re2
```
[the remark this answers is not in the document]
Comment (jcwchen) [reply onnx/reference/ops/op_regex_full_match.py:7]: +1. Could we also have a clear error message when re2 is not found? For instance, pyre2 is required to install for reference implementation of RegexFullMatch.

File: requirements-dev.txt (hunk 7 above)
Comment (adityagoel4512) [reply requirements-dev.txt:11]: If I'm not mistaken writing `google-re2[text]` in the requirements.txt file just means we request the google-re2 package with a nonexistent text option (we want the opposite).

File: onnx/defs/text/defs.cc (hunk 9 above)
Comment (adityagoel4512) [reply onnx/defs/text/defs.cc:33]: As we discussed, I don't think RE2 has explicit versioning unfortunately but the syntax looks very stable. It is also the syntax used by [tensorflow](https://www.tensorflow.org/api_docs/python/tf/strings/regex_full_match) and [PyTorch](https://pytorch.org/text/stable/transforms.html#regextokenizer) regex ops.

**Comment by adityagoel4512** [comment]
Updates:
- we lazily import re2
- `google-re2` is an optional dependency under the "reference" name (`pip install onnx[reference]` to install reference implementation exclusive dependencies). New dependencies exclusive to the reference implementation and not used elsewhere can be placed in requirements-reference.txt
- tests that use the reference implementation of RegexFullMatch are skipped when the platform in win32

**Comment by adityagoel4512** [comment]
Is there anything holding back this PR? @gramalingam @xadupre @jcwchen

File: onnx/test/test_backend_reference.py [hunk 11]
```diff
@@ -172,6 +173,11 @@ def run_node(cls, node, inputs, device=None, outputs_info=None, **kwargs):
 # The following tests fail due to discrepancies (small but still higher than 1e-7).
 backend_test.exclude("test_adam_multiple")  # 1e-2
 
+if sys.platform == "win32":
```
Comment (xadupre) [inline onnx/test/test_backend_reference.py:176]: Can you add a comment about why it is skipped on Windows?
